### PR TITLE
dta4-model-convergence-check.

### PR DESCRIPTION
**Pull request by tommorris168**
There are two places where the GitHub comparison makes it look like I have changed the condition on a conditional panel that was already there, but I haven't.

### CODE REVIEW

File: R/analysis_page.R [hunk 1]
```diff
@@ -470,6 +490,48 @@ AnalysisPageServer <- function(id, data) {
...
+      newData$true1 <- newData$TP
+      newData$true0 <- newData$TN 
+      newData$study <- 1:N
+      Y = reshape(newData, direction = "long", varying = list( c("n1" , "n0") , c( "true1","true0" ) ) ,
+                  timevar = "sens" , times = c(1,0) , v.names = c("n","true") ) 
+      Y = Y[order(Y$id),]  
+      Y$spec<- 1-Y$sens
+      MA_Y = glmer( formula = cbind(  true , n - true ) ~ 0 + sens + spec + (0+sens + spec|study),
+                    data = Y , family = binomial , nAGQ=1 , verbose=0 ) 
+      return(summary(MA_Y))
```
Comment (JanionNevill) [inline R/analysis_page.R:514]: The code in this block seems to be repeated many times in the server, can it be rationalised please? I'm keen to not bloat this app any further.

File: R/sensititvity_page.R [hunk 2]
```diff
@@ -499,6 +519,36 @@ SensitivityAnalysisPageServer <- function(id, data) {
...
+      newData$true1 <- newData$TP
+      newData$true0 <- newData$TN 
+      newData$study <- 1:N
+      Y = reshape(newData, direction = "long", varying = list( c("n1" , "n0") , c( "true1","true0" ) ) ,
+                  timevar = "sens" , times = c(1,0) , v.names = c("n","true") ) 
+      Y = Y[order(Y$id),]  
+      Y$spec<- 1-Y$sens
+      MA_Y = glmer( formula = cbind(  true , n - true ) ~ 0 + sens + spec + (0+sens + spec|study),
+                    data = Y , family = binomial , nAGQ=1 , verbose=0 ) 
+      return(summary(MA_Y))
```
Comment (JanionNevill) [inline R/sensititvity_page.R:543]: This seems to be the same code repeated again 😲 This might be a bigger job than I'd realised.

**Review by JanionNevill — CHANGES_REQUESTED** [review]
Code repetition needs addressing. I'll leave it to you to decide whether it's feasible on this task, but it needs doing before much more development is done on this app.

File: R/analysis_page.R [hunk 3]
```diff
@@ -470,6 +490,48 @@ AnalysisPageServer <- function(id, data) {
...
+    #Add converged() reactive to the output list
+    output$converged <- reactive(converged())
+    outputOptions(x = output, name = "converged", suspendWhenHidden = FALSE)
+    
+    #If the model did not converge, display a warning
+    observe({
+      if (!converged()) {
+        shinyalert(
+          title = "WARNING - model failed to converge",
+          text = paste0("Most output will not be displayed. \n\n MetaDTA is not equipped for model diagnostics. You are advised to use other software. \n\nError message: \n", model_summary()$optinfo$conv$lme4$messages),
```
Comment (JanionNevill) [inline R/analysis_page.R:529]: Is it worth pointing the user to MetaBayesDTA?

**Review by JanionNevill — CHANGES_REQUESTED** [review]
Sorry, additional comments I thought of after the first pass.

It might also be handy to show a message on each page saying something simple like "Analysis unavailable - Model did not converge". It might be best to stick this in a module to wrap around each page.

File: R/analysis_page.R (hunk 3 above)
Comment (tommorris168) [reply R/analysis_page.R:529]: Yes, I forgot about that.

**Comment by tommorris168** [comment]
I haven't specifically responded to the duplication you mentioned, but I have removed a lot of duplication. There is still a lot more than can be done, but I don't want to do it in this task, as the next step is harder.
There are a couple of things I'd be interested to hear your thought on:
1. For the sROC curve on line 495 of analysis_page.R I have created a function DrawRocPlot() rather than a reactive. This is because the download in output$downloadROC didn't work with a reactive. I don't know why.
2. Is it the right thing to have a ui function like the one in non_convergence_warning.R? I didn't add an id whenever I called it, so it seems like a ui might not be the right thing.

**Comment by tommorris168** [comment]
To elaborate on the duplicated code I have removed, whenever there was a display and an associated download, the object was created twice, once for the display and once for the download. I have created it once and then used it in both locations. In the worst example, the sroc curve had 1000 lines duplicated.
This can be called "within-tab" duplication. The next step would be to removed "between-tab" duplication.

File: R/analysis_page.R [hunk 4]
```diff
@@ -428,177 +474,65 @@ AnalysisPageServer <- function(id, data) {
...
-          title = "WARNING - model failed to converge",
-          text = paste0("Most output will not be displayed. \n\n MetaDTA is not equipped for model diagnostics. You are advised to use MetaBayesDTA (https://www.gla.ac.uk/research/az/crsu/apps-materials-guidence/) or other software. \n\nError message: \n", model_summary()$optinfo$conv$lme4$messages),
-          size = "l"
-        )
-      }
-    })
-    
     # Plot an interactive SROC curve 
-    output$sroc <- renderPlot({
+    DrawRocPlot <- function() {
```
Comment (JanionNevill) [inline R/analysis_page.R:495]: Can this function be extracted into a separate file please? It's best practice to pull out things which are not specific to the UI, for instance, this plot could be generated within a normal R script if the required inputs had been calculated.

**Comment by JanionNevill** [comment]
> 1. For the sROC curve on line 495 of analysis_page.R I have created a function DrawRocPlot() rather than a reactive. This is because the download in output$downloadROC didn't work with a reactive. I don't know why

I believe that this is due to how the plot commands are captured, but it is not very intuitive. Separating it out into a function is how I've done it in the past, so this looks good to me.

**Comment by JanionNevill** [comment]
> 2\. Is it the right thing to have a ui function like the one in non_convergence_warning.R? I didn't add an id whenever I called it, so it seems like a ui might not be the right thing.

It's not quite how I've done it, but this does work. The ID is not needed, as you say, because it's not actually a module, just a piece of UI code.

Have a look at [informed_conditional_panel.R](https://github.com/CRSU-Apps/MetaInsight/blob/feature-meta-regression/R/informed_conditional_panel.R) in MetaInsight. That achieves the same thing in a modular way. The code is more complex, but the conditional panels are dealt with inside the module, rather than on every tab.

File: R/analysis_page.R [hunk 5]
```diff
@@ -378,11 +376,59 @@ AnalysisPageServer <- function(id, data) {
...
+    #Add converged() reactive to the output list
+    output$converged <- reactive(converged())
+    outputOptions(x = output, name = "converged", suspendWhenHidden = FALSE)
+    
+    #If the model did not converge, display a warning
+    observe({
+      if (!converged()) {
+        shinyalert(
+          title = "WARNING - model failed to converge",
+          text = paste0("Most output will not be displayed. \n\n MetaDTA is not equipped for model diagnostics. You are advised to use MetaBayesDTA (https://www.gla.ac.uk/research/az/crsu/apps-materials-guidence/) or other software. \n\nError message: \n", model_summary()$optinfo$conv$lme4$messages),
```
Comment (JanionNevill) [inline R/analysis_page.R:399]: Can you make this a link please? Also, should it go straight to the app, or have you deliberately sent users to the website to show them all the other apps too?

**Review by JanionNevill — CHANGES_REQUESTED** [review]
Minor changes, and some comments.

File: R/analysis_page.R (hunk 5 above)
Comment (tommorris168) [reply R/analysis_page.R:399]: It's deliberate to send them to the "home page".

**Review by JanionNevill — APPROVED** [review]
✔

**Pull request merged**